File: LeetCode/102.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        result = list()
        if root:
            self.stacks = [[root], []]
            flag = 0
            while self.stacks[flag]:
                current_level = list()
                for node in self.stacks[flag]:
                    current_level.append(node.val)
                    if node.left:
                        self.stacks[1 - flag].append(node.left)
                    if node.right:
                        self.stacks[1 - flag].append(node.right)

                self.stacks[flag] = []
                result.append(current_level)
                flag = 1 - flag

        return result

File: Cpp/LeetCode/102.cpp
//
// Created by caizixiang on 2019-07-30.
//
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> order;
        if (!root)
            return order;

        queue<TreeNode*> current_layer;
        current_layer.push(root);

        while (!current_layer.empty())
        {
            vector<int> this_layer;
            queue<TreeNode*> next_layer;
            while (!current_layer.empty())
            {
                TreeNode* node = current_layer.front();
                current_layer.pop();
                this_layer.push_back(node->val);
                if (node->left)
                    next_layer.push(node->left);
                if (node->right)
                    next_layer.push(node->right);


            }

            order.push_back(this_layer);
            current_layer = next_layer;
        }

        return order;
    }
};